# Reading PDFs

## Install required libraries

In [ ]:
## install PyPDF2 as part of our process
## It's not part of the standard library.


In [ ]:
## Import libraries
## in order to export our file to our computer drive, you need this only in Colab:
# from google.colab import files


## Upload demo PDFs

In [ ]:
## import colab's file uploader
## upload 
# files.upload()

## 1. Reading fairly well-behaving non-tabular PDFs


In [ ]:
## read and store document in an object


In [ ]:
## type of object


In [ ]:
## use pypdf to read that object


In [ ]:
## type of object


In [ ]:
## let's see how many pages we have


In [ ]:
## read a single page 


In [ ]:
## store in an object first


In [ ]:
## read all pages and store in a list


In [ ]:
## LENGHT


In [ ]:
## print all pages so we can see them


In [ ]:
## Remember to close all open objects to avoid resource/processing power drain


In [ ]:
## more efficient way since we are not leaving file open


In [ ]:
## CALL 


## 2. Not always so lucky - troublesome PDFs

In [ ]:
## let's read nixon-memo2.pdf
## read and store document in an object


## 3. Obnoxious PDF

In [ ]:
## place "columbus_bank_trust.pdf" in an object


In [ ]:
## read it using PyPDF


In [ ]:
## how many pages?


In [ ]:
## extract a page


In [ ]:
## read text


In [ ]:
## remember to close


# Strategy to Vanquish Obnoxious PDFs




### The problem:
*   PDFs all have different encodings: UTF-8, ASCII, Unicode, etc
*   Therefore a possible loss of data during the conversion 

### The solution:
*   Convert the PDF to an image
*   Use optical character recognition (OCR) to capture the text
*   Export to a text file





### Let's use Google's tesseract OCR (and all its dependencies)

# Install and OCR for English documents

### Step 1. Create a virtual environment

Because Jupyter can't install these packages, we first need to create a virtual environment.

- Launch terminal and ```mkdir``` a folder.
- ```cd``` into that folder.
- type: ```Python3 -m venv venv``` to create a virtual environment
- type: ```source ./venv/bin/activate``` to activate that virtual environment

### Step 2. Install Tesseract

- We need Tesseract, Google's optical character recognition engine (not part of standard library). It can't be ```pip``` installed.

- In Terminal, in the virtual environment you created, type: ```brew install tesseract```

In [1]:
## What languages do we have installed now?
!tesseract --list-langs

List of available languages (163):
afr
amh
ara
asm
aze
aze_cyrl
bel
ben
bod
bos
bre
bul
cat
ceb
ces
chi_sim
chi_sim_vert
chi_tra
chi_tra_vert
chr
cos
cym
dan
deu
div
dzo
ell
eng
enm
epo
equ
est
eus
fao
fas
fil
fin
fra
frk
frm
fry
gla
gle
glg
grc
guj
hat
heb
hin
hrv
hun
hye
iku
ind
isl
ita
ita_old
jav
jpn
jpn_vert
kan
kat
kat_old
kaz
khm
kir
kmr
kor
kor_vert
lao
lat
lav
lit
ltz
mal
mar
mkd
mlt
mon
mri
msa
mya
nep
nld
nor
oci
ori
osd
pan
pol
por
pus
que
ron
rus
san
script/Arabic
script/Armenian
script/Bengali
script/Canadian_Aboriginal
script/Cherokee
script/Cyrillic
script/Devanagari
script/Ethiopic
script/Fraktur
script/Georgian
script/Greek
script/Gujarati
script/Gurmukhi
script/HanS
script/HanS_vert
script/HanT
script/HanT_vert
script/Hangul
script/Hangul_vert
script/Hebrew
script/Japanese
script/Japanese_vert
script/Kannada
script/Khmer
script/Lao
script/Latin
script/Malayalam
script/Myanmar
script/Oriya
script/Sinhala
script/Syriac
script/Tamil
script/Telugu
script/Thaana
script/Th

### Step 3. Install Poppler

- We need Poppler which converts ```.pdf```  to ```.txt``` files. But again, it is not part of the standard library and can't be pipped.
- In Terminal, in the same virtual environment, type: ```brew install poppler```.

### Step 4. Install Pytesserct

Pytesserct is a Python interface that allows us to work with Google's Tesseract.

- ```pip install``` it from PyPi (hosted by Python Software Foundation).

In [ ]:
#!pip install pytesseract


### Step 5. install pdf2image
This utility converts PDFs to PIL Image objects.

In [ ]:
## !pip install pdf2image


The Image module provides a class with the same name which is used to represent a PIL image. The module also provides a number of factory functions, including functions to load images from files, and to create new images. (credit: Image Module)

In [ ]:
# Import libraries
from PIL import Image
import pytesseract
import os ## navitage, create directories
import shutil ## to delete the image folders with their imgs
from pdf2image import convert_from_path ## to turn pdf to image
import glob ## to glob files into a list
from pathlib import Path ## to specify path to your files
from natsort import natsorted, ns ## natural sorting
## in order to export our file to our computer drive, you need this only in Colab:
# from google.colab import files

In [ ]:
## import colab's file uploader
## upload 
# files.upload()

In [ ]:
## path to nixon-memo2.pdf


In [ ]:
pages = convert_from_path(PDF_file, 300) ## 500 dpi. balance between hi-res and computation power
print(type(pages))
pages

In [ ]:
## Take your PDF and convert each page to a JPEG

# Store all the pages of the PDF in a variable
pages = convert_from_path(PDF_file, 300) ## 500 dpi. balance between hi-res and computation power
print(pages)
print(f"Give me a second...converting '{PDF_file}' to a JPEG")

# Counter to store images of each page of PDF to image
image_counter = 1

# Iterate through all the pages stored above
for page in pages:
    ## Declaring filename for each page of PDF as JPG
    ## remove the .pdf extention (last 4 characters)
    filename = str(PDF_file[ : -4])+"_page_"+str(image_counter)+".jpg"
    print(f"Here we go: {filename}")
    print(type(filename))
    # Save the image of the page in system
    page.save(filename, 'JPEG') ## here we tell it to be a "jpeg" file
    # Increment the counter to update filename
    image_counter+=1

In [ ]:
## path to the jpegs that were just produced

img_path = "nixon*.jpg"
myfiles = sorted(glob.glob(img_path))
myfiles

In [ ]:
## name your exported file
outfile = "nixon.txt"

In [ ]:
## Open each Jpeg, OCR and append text to output file

with open(outfile, "a") as my_conversion_file:
  ## grab all jpegs
  img_files = sorted(glob.glob(img_path))
  print(img_files)
  counter = 0
    ## name each file
  for img_file in img_files:
    counter+=1
    print(f"Converting image {counter} of {len(img_files)} to text")
    text = f"FILE_Info: {img_file} \n\n\n + {str(pytesseract.image_to_string(Image.open(img_file)))}" 

  ## In many PDFs, at line ending, if a word can't
    ## be written fully, a 'hyphen' is added.
    ## To remove this, we replace every '-\n' to ''.
    text = text.replace('-\n', '')

    my_conversion_file.write(text)

print(f"Your converted PDFs are waiting in {outfile}!")

## Turn into Functions

In [ ]:
### FUNCTION to convert pdf to jpg
def to_jpg(PDF_file, pdf_path, resolution = 300):
    print("Getting all your PDFs from target folder...")
    ## store in a folder for conversion
    conversion_path = Path('conversion_img/')
    conversion_path.mkdir(exist_ok=True)

    '''
    Part #1 : Converting PDF to images
    '''

    # Store all the pages of the PDF in a variable
    pages = convert_from_path(pdf_path+PDF_file, resolution)
    print(f"Give me a second...converting '{PDF_file}' to a JPEG")

    # Counter to store images of each page of PDF to image
    image_counter = 1

    # Iterate through Sall the pages stored above
    for page in pages:
        ## Declaring filename for each page of PDF as JPG
        ## remove the .pdf extention (last 4 characters)
        filename = str(PDF_file[ : -4])+"_page_"+str(image_counter)+".jpg"
        print(f"Here we go: {filename}")
        filename = os.path.join(conversion_path,filename)
        print(filename)
        # Save the image of the page in system
        page.save(filename, 'JPEG')

        # Increment the counter to update filename
        image_counter+=1

In [ ]:
### FUNCTION TO OCR JPG

def jpg_to_txt(outfile):

    '''
    Part #2 - Recognizing text from the images using OCR
    '''
    ## store in an output folder
    output_path = Path('output/')
    output_path.mkdir(exist_ok=True)
    conversion_path = Path('conversion_img/')

    # # # # Creating a text file to write the output
    outfile = os.path.join(output_path,outfile)
    # # #
    # # # # Open output file and append all info

    with open(outfile, "a") as f:
        ## grab all jpegs
        img_files = sorted(Path(conversion_path).glob('*.jpg'), reverse=False)

        ## path objects turn urls into strings and you can't sort numbers with underscores, etc
        ## the function below uses regex to find the numbers and turn them into integer.
        ## natsort library
        ## https://natsort.readthedocs.io/en/master/howitworks.html
        from natsort import natsorted, ns
        img_files = [img_file.as_posix() for img_file in img_files]
        img_files = natsorted(img_files, alg=ns.PATH)
        print(img_files)

        counter = 0
        for img_file in img_files:
            counter+=1
            print(f"Converting image {counter} of {len(img_files)} to text!")
            text = f"\n\n\nFILE_Info: {img_file} \n\n\n \
             {str(pytesseract.image_to_string(Image.open(img_file)))}"
            # ###UNCHECK IF SPANISH
            # text = f"\n\n\nFILE_Info: {img_file} \n\n\n + {str(pytesseract.image_to_string(Image.open(img_file), lang='spa'))}"
            # print(text)

            ## In many PDFs, at line ending, if a word can't
            ## be written fully, a 'hyphen' is added.
            ## To remove this, we replace every '-\n' to ''.
            text = text.replace('-\n', '')
            # print(f"Here's the text: \n {text}")

            f.write(text)

            ## Delete the folder
        shutil.rmtree(conversion_path)
        print(f"A text file called '{outfile}' is ready for you....")

In [ ]:
def ocr_file(PDF_file, outfile, pdf_path = "", resolution = 300 ):
    '''
    args:
    - pdf file name(as strings)
    - name of outfile (as strings)
    - path to pdf as string (default "" same folder)
    - resolution as int between 300 and 600
    '''
    to_jpg(PDF_file, pdf_path, resolution)
    jpg_to_txt(outfile)
    

In [ ]:
## ocr the file "nixon-memo1.pdf"


In [ ]:
## ocr the file "bank.pdf"


# OCRing languages other than English.

English is the default language for Tesseract. 

The installation process is confusing and cumbersome. I have have tried to make it as clear as possible here in *Colab*. 

Installation in Jupyter Notebooks is a little different and so is installation for a virtual environment.

[More here](https://askubuntu.com/questions/793634/how-do-i-install-a-new-language-pack-for-tesseract-on-16-04
).



# Installations

### Step 1. Install languages

- ```brew install tesseract-lang``` 

Note: This will install about 163 languages. The method to install specific languages is beyond the scope of this class.

In [ ]:
## What languages do we have installed so far?
!tesseract --list-langs

In [ ]:
## What languages do we have installed now?
!tesseract --list-langs

In [ ]:
## import colab's file uploader
## upload 
# files.upload()


In [ ]:
## grab 'spanish.pdf'
PDF_file = "spanish.pdf"


In [ ]:
## Take your PDF and convert each page to a JPEG

# Store all the pages of the PDF in a variable
pages = convert_from_path(PDF_file, 500) ## 500 dpi. balance between hi-res and computation power
print(pages)
print(f"Give me a second...converting '{PDF_file}' to a JPEG")

# Counter to store images of each page of PDF to image
image_counter = 1

# Iterate through all the pages stored above
for page in pages:
    ## Declaring filename for each page of PDF as JPG
    ## remove the .pdf extention (last 4 characters)
    filename = str(PDF_file[ : -4])+"_page_"+str(image_counter)+".jpg"
    print(f"Here we go: {filename}")

    # Save the image of the page in system
    page.save(filename, 'JPEG')

    # Increment the counter to update filename
    image_counter+=1

In [ ]:
## path to the jpegs that were just produced
img_path = "spanish*.jpg"
myfiles = sorted(glob.glob(img_path))
myfiles

In [ ]:
## The text pulled from the jpegs will be appended to this file
outfile = "spanish_TEXT.txt"

In [ ]:
## Open each Jpeg, OCR and append text to output file

##REMEMBER TO CHANGE THE LANGUAGE

with open(outfile, "a") as my_convertion_file:
    ## grab all jpegs
    img_files = sorted(glob.glob(img_path))
    counter = 0
    for img_file in img_files:
        counter+=1
        print(f"Converting image {counter} of {len(img_files)} to text!")
        # text = f"\n\n\nFILE_Info: {img_file} \n\n\n + {str(pytesseract.image_to_string(Image.open(img_file)))}" 
        
        # ###UNCHECK for Arabic
        text = f"\n\n\nFILE_Info: {img_file} \n\n\n {str(pytesseract.image_to_string(Image.open(img_file), lang='spa'))}"
        print(text)

        ## In many PDFs, at line ending, if a word can't
        ## be written fully, a 'hyphen' is added.
        ## To remove this, we replace every '-\n' to ''.
        text = text.replace('-\n', '')
        # print(f"Here's the text: \n {text}")

        my_convertion_file.write(text) ## write to my output file

# Let's try Chinese

In [ ]:
## What languages do we have installed so far?
!tesseract --list-langs

In [ ]:
!apt-get install tesseract-ocr-chi-tra

In [ ]:
## import colab's file uploader
## upload 
files.upload()

In [ ]:
PDF_file = "chinese.pdf"

In [ ]:
## Take your PDF and convert each page to a JPEG

# Store all the pages of the PDF in a variable
pages = convert_from_path(PDF_file, 500) ## 500 dpi. balance between hi-res and computation power
print(pages)
print(f"Give me a second...converting '{PDF_file}' to a JPEG")

# Counter to store images of each page of PDF to image
image_counter = 1

# Iterate through all the pages stored above
for page in pages:
    ## Declaring filename for each page of PDF as JPG
    ## remove the .pdf extention (last 4 characters)
    filename = str(PDF_file[ : -4])+"_page_"+str(image_counter)+".jpg"
    print(f"Here we go: {filename}")

    # Save the image of the page in system
    page.save(filename, 'JPEG')

    # Increment the counter to update filename
    image_counter+=1

In [ ]:
## path to the jpegs that were just produced
img_path = "chinese*.jpg"
myfiles = sorted(glob.glob(img_path))
myfiles

In [ ]:
## The text pulled from the jpegs will be appended to this file
outfile = "chinese_TEXT.txt"

In [ ]:
## Open each Jpeg, OCR and append text to output file

with open(outfile, "a") as my_convertion_file:
    ## grab all jpegs
    img_files = sorted(glob.glob(img_path))
    counter = 0
    for img_file in img_files:
        counter+=1
        print(f"Converting image {counter} of {len(img_files)} to text!")
        # text = f"\n\n\nFILE_Info: {img_file} \n\n\n + {str(pytesseract.image_to_string(Image.open(img_file)))}" 
        
        # ###UNCHECK for Arabic
        text = f"\n\n\nFILE_Info: {img_file} \n\n\n  {str(pytesseract.image_to_string(Image.open(img_file), lang='chi_tra'))}"
        print(text)

        ## In many PDFs, at line ending, if a word can't
        ## be written fully, a 'hyphen' is added.
        ## To remove this, we replace every '-\n' to ''.
        text = text.replace('-\n', '')
        # print(f"Here's the text: \n {text}")

        my_convertion_file.write(text) ## write to my output file

# Let's try Japanese


In [ ]:
## get Japanese OCR
!apt-get install tesseract-ocr-jpn

In [ ]:
## What languages do we have installed so far?
!tesseract --list-langs

In [ ]:
## import colab's file uploader
## upload 
files.upload()

In [ ]:
PDF_file = "japanese.pdf"


In [ ]:
## Take your PDF and convert each page to a JPEG

# Store all the pages of the PDF in a variable
pages = convert_from_path(PDF_file, 500) ## 1000 dpi. balance between hi-res and computation power
print(pages)
print(f"Give me a second...converting '{PDF_file}' to a JPEG")

# Counter to store images of each page of PDF to image
image_counter = 1

# Iterate through all the pages stored above
for page in pages:
    ## Declaring filename for each page of PDF as JPG
    ## remove the .pdf extention (last 4 characters)
    filename = str(PDF_file[ : -4])+"_page_"+str(image_counter)+".jpg"
    print(f"Here we go: {filename}")

    # Save the image of the page in system
    page.save(filename, 'JPEG')

    # Increment the counter to update filename
    image_counter+=1

In [ ]:
## path to the jpegs that were just produced
img_path = "japanese*.jpg"
myfiles = sorted(glob.glob(img_path))
myfiles

In [ ]:
## The text pulled from the jpegs will be appended to this file
outfile = "japanese_TEXT.txt"

In [ ]:
## Open each Jpeg, OCR and append text to output file

with open(outfile, "a") as my_convertion_file:
    ## grab all jpegs
    img_files = sorted(glob.glob(img_path))
    counter = 0
    for img_file in img_files:
        counter+=1
        print(f"Converting image {counter} of {len(img_files)} to text!")
        # text = f"\n\n\nFILE_Info: {img_file} \n\n\n + {str(pytesseract.image_to_string(Image.open(img_file)))}" 
        
        # ###UNCHECK for Arabic
        text = f"\n\n\nFILE_Info: {img_file} \n\n\n  {str(pytesseract.image_to_string(Image.open(img_file), lang='jpn'))}"
        print(text)

        ## In many PDFs, at line ending, if a word can't
        ## be written fully, a 'hyphen' is added.
        ## To remove this, we replace every '-\n' to ''.
        text = text.replace('-\n', '')
        # print(f"Here's the text: \n {text}")

        my_convertion_file.write(text) ## write to my output file